config/decoder.go

```diff
@@ -294,6 +295,41 @@ func (d *decoder) pointer(name string, value reflect.Value, def string) error {
 	return d.unmarshal(name, value.Elem(), def)
 }
 
+// Sets value to a channel type.
+func (d *decoder) channel(key string, value reflect.Value, def string) error {
+	return d.textUnmarshaller(key, value, def)
+}
+
+// Sets value to a function type.
+func (d *decoder) function(key string, value reflect.Value, def string) error {
+	return d.textUnmarshaller(key, value, def)
+}
+
+func (d *decoder) textUnmarshaller(key string, value reflect.Value, str string) error {
+	v := d.getGlobalProvider().Get(key)
+	if v.HasValue() {
+		str = v.String()
+	} else if str == "" {
+		return nil
+	}
+
+	if value.IsNil() {
+		value.Set(reflect.New(value.Type()).Elem())
+	}
+
+	// Value has to have a pointer receiver to be able to modify itself with TextUnmarshaller
+	if !value.CanAddr() {
+		return fmt.Errorf("can't use TextUnmarshaller because %s is not addressable", key)
+	}
+
+	switch t := value.Addr().Interface().(type) {
+	case encoding.TextUnmarshaler:
+		return t.UnmarshalText([]byte(str))
```

Comment:
Naive question: is any of this code going to be used in the "hot path"? Allocating a new byte slice may be small but non-negligable, same question about all the reflect stuff. I'm far enough away from the code right now that it's unclear to me how deep we're digging here.

IOW, do we need to add this to our benchmarks?